Add Counterstrike mission Siberian Conflict 1: Fresh Tracks

Smittytron opened the pull request.

Smittytron committed:
Add Counterstrike mission Siberian Conflict 1: Fresh Tracks

Mailaender commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/rules.yaml:
```diff
@@ -0,0 +1,71 @@
+
+HELI:
+	Buildable:
+		Prerequisites: ~disabled
```
Is this intentional?

Smittytron replied:
Yes, the starting helicopter was not replaceable in this mission.

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+	if ticked > 0 then
+		UserInterface.SetMissionText("First trucks arrive in " .. Utils.FormatTime(ticked), TimerColor)
+		ticked = ticked - 1
+	elseif ticked == 0 and not ConvoysSent == true then
```
`not ConvoysSent then`

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+	ConvoysPassed = ConvoysPassed + 1
+	if ConvoysPassed <= ConvoyWaves[Map.LobbyOption("difficulty")] then
+		Trigger.AfterDelay(DateTime.Seconds(90), SendConvoys)
+			Media.PlaySpeechNotification(allies, "ConvoyApproaching")
```
The indentation here looks wrong.

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+
+OpeningMoves = function()
+	Utils.Do(Squad1, function(actor)
+		if not actor.IsDead then
```
These actors surely can't be dead when the world is loaded?

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+	end)
+
+	Trigger.AfterDelay(Utils.RandomInteger(AttackWaveDelay[1], AttackWaveDelay[2]), function()
+		AttackWaves()
```
We can simplify this to `Trigger.AfterDelay(Utils.RandomInteger(AttackWaveDelay[1], AttackWaveDelay[2]), AttackWaves)`

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+		end)
+	end)
+
+	Trigger.AfterDelay(DateTime.Seconds(127), function()
```
This again looks weird. Why don't you use the `finalConvoy` table in the trigger above to attach the `OnAllKilled` trigger to?

Smittytron replied:
Just the trucks need to be killed instead of the whole convoy. I guess I could split it up into a separate group of reinforcements that are just trucks. Also, it is theoretically possible, though silly, that a player could stop a truck from an earlier convoy with non-combatants, which is why I chose to `GetActors` here.

abcdefg30 replied:
I was thinking of something like
```
Trigger.OnAllKilled(Utils.Where(finalConvoy, function(a) return a.Type == "truk" end), function()
	Allies.MarkCompletedObjective(StopTrucks)
end)
```

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+
+	ussrObj = ussr.AddPrimaryObjective("Defeat the Allies.")
+	StopTrucks = allies.AddPrimaryObjective("Destroy all Soviet convoy trucks.")
+	DestroyBridges = allies.AddSecondaryObjective("Destroy the nearby bridges to slow the  \nconvoys down.")
```
Style nit: No spaces before a `\n`.

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/rules.yaml:
```diff
@@ -0,0 +1,71 @@
+
+HELI:
+	Buildable:
+		Prerequisites: ~disabled
```
Considering how the enemy has flack trucks in our version (and the original only had mammoth which didn't do too much against air), we should rethink enabling this.

Smittytron replied:
Yeah that makes sense. Also nukes are brutal vs air even at the outer limits.

abcdefg30 reviewed (commented):
The map doesn't lock the preview image.

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,275 @@
+	Trigger.AfterDelay(DateTime.Minutes(2), function()
+		Media.PlaySpeechNotification(Allies, "ConvoyApproaching")
+		local finalConvoy = Reinforcements.Reinforce(USSR, FinalConvoyUnits, { NorthEntry.Location })
+		Utils.Do(FinalConvoyPath, function(waypoint)
```
Why isn't this the path given to `Reinforcements.Reinforce`? Because units should not attack-move? Either way, we're currently adding on `OnAddedToWorld` trigger for every waypoint.
Better would be (if we don't give the path to `Reinforcements.Reinforce`):
```
Utils.Do(finalConvoy, function(actor)
	Trigger.OnAddedToWorld(actor, function(a)
		Utils.Do(FinalConvoyPath, function(waypoint)
			a.Move(waypoint.Location)
		end)
	end)
end)
```

Smittytron commented:
Updated

phrohdoh commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+Squad1 = { FirstSquad1, FirstSquad2, FirstSquad3 }
+Squad2 = { SecondSquad1, SecondSquad2, SecondSquad3 }
+PatrolMammothPath = { Patrol1.Location, Patrol2.Location, Patrol3.Location, Patrol4.Location, Patrol5.Location, Patrol6.Location, Patrol7.Location }
+ConvoyEscape = { CPos.New(113, 42), CPos.New(117, 71) }
```
Is this generally how this is done, or do we usually place `waypoint`s (and use their `Location`)?

abcdefg30 replied:
For cell triggers we usually do it this way.

Smittytron commented:
I went ahead and switched the crate explosion from Atomic to MiniNuke.

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+	Utils.Do(attackers, function(unit)
+		Trigger.OnAddedToWorld(unit, function()
+			unit.AttackMove(AlliedBase.Location)
+			unit.Hunt()
```
Shouldn't this use `IdleHunt`?

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+			c = HSLColor.White
+		end
+
+		Trigger.AfterDelay(DateTime.Seconds(i), function() UserInterface.SetMissionText("Here they come!", c) end)
```
Imo it would be better to have a more formal tone in the UI. What about just "The first convoy has entered the area"?

Smittytron replied:
How about: "The first trucks are entering your AO."

abcdefg30 replied:
That should be fine.

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/rules.yaml:
```diff
@@ -0,0 +1,71 @@
+	MissionData:
+		WinVideo: sovbatl.vqa
+		LossVideo: overrun.vqa
+		Briefing: Soviet activity in Siberia is increasing. A large shipment of atomic material is on its way through this area, and we need it stopped. Destroy all of the Soviet convoy trucks and prevent the Soviets from furthering their atomic testing.\n\nBe careful of the convoy truck cargo -- atomic material is highly unstable.\n\nThe mission is complete when all the convoy trucks from each of the  ...
```
This will be a problem: The briefing explicitly says "seven convoys". But it should be easy to resolve by just removing "seven" here (the sentence is still ok without it imo).

Smittytron commented:
Updated

Smittytron commented:
Updated for feedback from @Orb370. Crate sprites changed to wooden for better contrast. Timer reduced by 2 minutes. Added a second ore refinery because ore trucks will go idle when past the three per refinery limit.

Smittytron commented:
Rebased

abcdefg30 commented on mods/ra/maps/siberian-conflict-1-fresh-tracks/fresh-tracks.lua:
```diff
@@ -0,0 +1,282 @@
+	Trigger.AfterDelay(DateTime.Minutes(7), function()
+		Media.PlaySpeechNotification(Allies, "WarningThreeMinutesRemaining")
+	end)
+	Trigger.AfterDelay(DateTime.Minutes(9), function()
```
This needs to be adjusted for the shorter timer.

Smittytron commented:
Fixed timer notifications

abcdefg30 commented:
Lgtm and merging with @Orb370's second +1.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f4399ab4d4c66fe6fba5286082a6b43aacddcf6f...2f087b27dedfbe636e0cb391da1974fd9708b346)
